src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -654,18 +673,87 @@ def _map_line_buffers(
         point_buffer.append(indent_point)
 
         if not indent_point.is_line_break:
+            # If it's not a line break, we should still check whether it's
+            # untaken to keep track of them.
+            if indent_point.indent_impulse:
+                untaken_indent_locs[
+                    indent_point.initial_indent_balance + indent_point.indent_impulse
+                ] = indent_point.idx
             continue
 
         # If it *is* a line break, then store it.
         lines.append(_IndentLine.from_points(point_buffer))
+
+        # We should also evaluate whether this point inserts a newline at the close
+        # of an indent which was untaken on the way up.
+        # https://github.com/sqlfluff/sqlfluff/issues/4234
+        # Special case 1:
+        # If we're at the end of the file we shouldn't interpret it as a line break
+        # for problem indents, they're a bit of a special case.
+        # Special case 2:
+        # Bracketed expressions are a bit odd here.
+        # e.g.
+        #   WHERE (
+        #       foo = bar
+        #   )
+        #   LIMIT 1
+        #
+        # Technically there's an untaken indent before the opening bracket
+        # but this layout is common practice so we're not going to force
+        # one there even though there _is_ a line break after the closing
+        # bracket.
+        following_class_types = elements[indent_point.idx + 1].class_types
+        if (
+            indent_point.indent_trough
+            # End of file ends case. (Special case 1)
+            and "end_of_file" not in following_class_types
+        ):
+            passing_indents = list(
+                range(
+                    indent_point.initial_indent_balance,
+                    indent_point.initial_indent_balance + indent_point.indent_trough,
+                    -1,
+                )
+            )
+            # For it to be a problem - all passing indents must be untaken. If any
+            # were taken, then we can interpret the drop as closing that one.
+            if all(i in indent_point.untaken_indents for i in passing_indents):
+                for i in passing_indents:
+                    # If we don't have the location of the untaken indent, then
+                    # skip it for now. TODO: Check this isn't a bug when this happens.
+                    # It seems very rare for now.
+                    if i not in untaken_indent_locs:
+                        continue
+
+                    loc = untaken_indent_locs[i]
+
+                    # First check for bracket special case. It's less about whether
+                    # the section _ends_ with a lone bracket, and more about whether
+                    # the _starting point_ is a bracket which closes a line. If it
+                    # is, then skip this location. (Special case 2)
+                    if "start_bracket" in elements[loc + 1].class_types:
```

Comment:
Possible `IndexError`?

Reply:
Same as above - we know that files end with a block, and we're in a point so I think we're safe. I'll add a note.